tests/test_hybrid.py

```diff
@@ -278,3 +278,74 @@ def test_sap_fem_vs_robot(show_viewer):
 
     # Check that the legs of the ants are resting on the sphere
     assert_allclose(robot.get_qpos()[-4:].abs(), 1.0, tol=0.1)
+
+
+@pytest.mark.required
+@pytest.mark.parametrize("substeps", [1, 10])
+def test_rigid_mpm_legacy_coupling(substeps, show_viewer):
+    # This test is aimining at two things:
+    # 1) When substep = 1, the rigid object should be affected by the MPM particles
+    # 2) Regardless of substeps, as far as the total step dt is the same, they should give consistent results
+    dt_per_substep = 4e-4
+    horizon_substeps = 2000
+    atol = 1e-2
+
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(
+            dt=dt_per_substep * substeps,
+            substeps=substeps,
+        ),
+        mpm_options=gs.options.MPMOptions(
+            lower_bound=(-0.5, -1.0, 0.0),
+            upper_bound=(0.5, 1.0, 1),
+        ),
+        vis_options=gs.options.VisOptions(
+            visualize_mpm_boundary=True,
+        ),
+        viewer_options=gs.options.ViewerOptions(
+            camera_fov=30,
+        ),
+        show_viewer=show_viewer,
+    )
+
+    plane = scene.add_entity(
+        morph=gs.morphs.Plane(),
+    )
+
+    obj_rigid = scene.add_entity(
+        material=gs.materials.Rigid(
+            rho=200.0,
+        ),
+        morph=gs.morphs.Box(
+            pos=(0.0, -0.5, 0.25),
+            size=(0.2, 0.2, 0.2),
+        ),
+        surface=gs.surfaces.Default(
+            color=(1.0, 0.4, 0.4),
+            vis_mode="visual",
+        ),
+    )
+
+    obj_sand = scene.add_entity(
+        material=gs.materials.MPM.Liquid(),
+        morph=gs.morphs.Box(
+            pos=(0.0, 0.0, 0.25),
+            size=(0.3, 0.3, 0.3),
+        ),
+        surface=gs.surfaces.Default(
+            color=(0.3, 0.3, 1.0),
+            vis_mode="particle",
+        ),
+    )
+
+    scene.build()
+
+    horizon = horizon_substeps // substeps
+    for i in range(horizon):
+        scene.step()
+
+    # Check that the sand moved the box to the correct position
+    pos = obj_rigid.get_pos()
+    target_pos = np.array([0.0, -0.55, 0.1])
```

Comment:
why is this target_pos the expected pos? Does this mean the sand push the objects in the -y direction?

Reply:
It's based on experiments, there is no closed form formulation to get this expected pos. Would it be better to only check if the box has been moved or not, rather than using this expected pos?